File: 496-next-greater-element-i/next-greater-element-i.py
class Solution:
    def nextGreaterElement(self, nums1: List[int], nums2: List[int]) -> List[int]:
        # Dictionary to hold the next greater element for each number in nums2
        next_greater = {}
        # Stack to keep track of the numbers for which we haven't found the next greater element yet
        stack = []

        # Traverse nums2 to fill the next_greater dictionary
        for num in nums2:
            # While stack is not empty and the current number is greater than the stack's top element
            while stack and stack[-1] < num:
                # Pop the stack and set the current number as the next greater element for the popped number
                next_greater[stack.pop()] = num
            # Push the current number onto the stack
            stack.append(num)

        # For any remaining numbers in the stack, there is no greater element, so set it to -1
        while stack:
            next_greater[stack.pop()] = -1

        # Construct the result list for nums1 using the next_greater dictionary
        return [next_greater.get(num, -1) for num in nums1]

File: 496-next-greater-element-i/next-greater-element-i.java
class Solution {
    public int[] nextGreaterElement(int[] nums1, int[] nums2) {
        int [] arr = new int[100000];

        Stack<Integer> stk = new Stack<>();

        for(int i = nums2.length-1;i>=0;i--){
            
            while(!stk.isEmpty() && stk.peek() < nums2[i]){
                stk.pop();
            }

            if(stk.isEmpty()){
                arr[nums2[i]] = -1;
            }
            else{
                arr[nums2[i]] = stk.peek();
            }

            stk.push(nums2[i]);

        }
        
        for(int i = 0; i < nums1.length ; i++){
            nums1[i] = arr[nums1[i]];
        }
       return nums1;
    }
}